ocaml/xapi/repository_helpers.ml

```diff
@@ -364,51 +370,55 @@ module UpdateInfoMetaData = struct
     ; location: string
   }
 
+  let assert_valid_repomd = function
+    | { checksum = ""; _ } | { location = ""; _ } ->
+      error "Missing 'checksum' or 'location' in updateinfo in repomod.xml";
+      raise Api_errors.(Server_error (invalid_repomd_xml, []))
+    | umd -> umd
+
+  let of_xml = function
+    | Xml.Element ("repomd", _, children) ->
```

Comment:
I believe this code could be made simpler using monadic error handling. For the success case, all expected information needs to be found and if any of that is missing, the fails with an error message. However, this needs some thoughts and I can't just write it down here myself.

Reply:
How about following?
`  let of_xml xml =
    let open Rresult.R.Infix in
    Ok xml
    >>= (function
        | Xml.Element ("repomd", _, children) ->
          let get_updateinfo_node = function
            | Xml.Element ("data", attrs, nodes) ->
              (match List.assoc_opt "type" attrs with
               | Some "updateinfo" -> Some nodes
               | _ -> None)
            | _ -> None
          in
          Ok (List.filter_map get_updateinfo_node children)
        | _ ->
          Error "Missing 'repomd' node")
    >>= (function
        | [l] ->
          Ok (List.fold_left (fun md n ->
              match n with
              | Xml.Element ("checksum", _, [Xml.PCData v]) ->
                {md with checksum = v}
              | Xml.Element ("location", attrs, _) ->
                (try {md with location = (List.assoc "href" attrs)}
                 with _ -> md)
              | _ -> md
            ) { checksum = ""; location = "" } l)
        | _ -> Error "Missing or multiple 'updateinfo' node(s)")
    >>= (function
        | { checksum = ""; _ } | { location = ""; _ } ->
          Error "Missing 'checksum' or 'location' in updateinfo in repomod.xml"
        | umd -> Ok umd)`

Reply:
How about following?
``` 
 let of_xml xml =
    let open Rresult.R.Infix in
    Ok xml
    >>= (function
        | Xml.Element ("repomd", _, children) ->
          let get_updateinfo_node = function
            | Xml.Element ("data", attrs, nodes) ->
              (match List.assoc_opt "type" attrs with
               | Some "updateinfo" -> Some nodes
               | _ -> None)
            | _ -> None
          in
          Ok (List.filter_map get_updateinfo_node children)
        | _ ->
          Error "Missing 'repomd' node")
    >>= (function
        | [l] ->
          Ok (List.fold_left (fun md n ->
              match n with
              | Xml.Element ("checksum", _, [Xml.PCData v]) ->
                {md with checksum = v}
              | Xml.Element ("location", attrs, _) ->
                (try {md with location = (List.assoc "href" attrs)}
                 with _ -> md)
              | _ -> md
            ) { checksum = ""; location = "" } l)
        | _ -> Error "Missing or multiple 'updateinfo' node(s)")
    >>= (function
        | { checksum = ""; _ } | { location = ""; _ } ->
          Error "Missing 'checksum' or 'location' in updateinfo in repomod.xml"
        | umd -> Ok umd)```